src/controllers/focus-lost-controller.js

```diff
@@ -0,0 +1,71 @@
+/**
+ * @callback FocusLostCallback
+ * @param {FocusLostEvent} event
+ * @returns {void}
+ */
+
+/**
+ * @typedef {Object} FocusLostEvent
+ * @property {Element} removedElement the element that has been removed from DOM
+ * @property {Element} focusedElement the element that had focus just before removal (can be the removedElement itself or one of its children)
+ * @property {number} index the index of the removedElement (before removal)
+ * @property {Element|null} suggestedElement the element that we suggest to gain focus
+ */
+
+import { findActiveElement, isParentOf } from '../lib/shadow-dom-utils.js';
+
+/**
+ * This Reactive controller helps detect a focus lost situation that can occur when the focused element is removed from the DOM.
+ *
+ * You provide a query selector which will select the elements for which you're tracking removal.
+ *
+ * You provide a callback that will be fired when one of these elements is removed from the DOM.
+ * The callback is fired only if the deleted element had focus (or one of its children had focus)
+ *
+ * The callback will suggest a new element to focus. It suggests the nearest element matching the same query selector.
+ */
+export class FocusLostController {
+  /**
+   *
+   * @param {LitElement} host the custom element
+   * @param {string} selector the query selector that will select the elements we want to listen to
+   * @param {FocusLostCallback} callback
+   */
+  constructor (host, selector, callback) {
+    host.addController(this);
+    this.host = host;
+    this.selector = selector;
+    this.elements = [];
+    this.callback = callback;
+  }
+
+  hostUpdate () {
+    // memorize the active element so that we are able to check if a deleted element was a parent of the focused element.
+    this.activeElement = findActiveElement();
+  }
+
+  hostUpdated () {
+    // keep the previous elements because we will want to compare with the new ones
+    const previousElements = this.elements;
+    // get the elements we are interested in
+    this.elements = [...this.host.shadowRoot.querySelectorAll(this.selector)];
+
+    // compare elements to the previous one and check if some elements have been removed
+    const removedElements = previousElements.filter((e) => !this.elements.includes(e));
+    if (removedElements.length > 0) {
+      // among these removed elements, find the one containing the focused element
+      // elementWithFocus may not be the focused element itself, it can be the element matching the query selector containing the focused element.
+      const elementWithFocus = removedElements.find((e) => e === this.activeElement || isParentOf(e, this.activeElement));
```

Comment:
I don't understand how this works:
we are using the `hostUpdated` hook.
This means we are running this after the new rendering phase is complete (before the `updated` lifecycle, but still, after the `render` phase).
At this point, the element we clicked has been removed and is no longer part of the DOM. The focus should already be lost (meaning, the `activeElement` can be determined by the browser, which is unreliable).

I don't understand how an element that is no longer part of the DOM could still be the `activeElement` ?
(I do see how it could be the parent of the activeElement though, depending on what the browser does when the focus is lost)

Reply:
Ahh I got it sorry, we get the `activeElement` during the `update` phase, gg :smile: